### ISSUE TITLE
Trigger a custom rendering

### ISSUE THREAD

**Issue by LeonardoGentile** [mobxjs/mobx-react#230]
Hello, 
right now by decorating any component with `@observer` will trigger a render call whenever any observable data (used inside the render method) changes.

What if we want to fine-tune when the render should be triggered? An example of this would be when creating a HOC for routing. I want to re-render a component and all its children only when a we are on a certain route/state (the routes changes coming from an observable).

In this case we don't use/need the observable inside the render method. Also we don't want to re-trigger the render for any route changes (only for specific ones).  

So this is what I've accomplished so far:

```
import React, { Component, createElement } from 'react';
import { autorun } from 'mobx';
import { inject} from 'mobx-react';

function routeNode(nodeName) { // route node Name
  return function routeNodeWrapper(RouteSegment) { // component Name

    @inject('routerStore')
    class RouteNode extends Component {

      constructor(props, context) {
        super(props, context);
        this.state = {
          route: props.routerStore.route,
          intersectionNode: props.routerStore.intersectionNode,
        };
      }

      componentDidMount() {
        this.autorunDisposer = autorun(() => {
          this.setState({
            route: this.props.routerStore.route,
            intersectionNode: this.props.routerStore.intersectionNode
          });
        });
      }

      componentWillUnmount() {
        this.autorunDisposer();
      }
      // Re-render the route-node (wrapped component) only if
      // it is the correct "transition node"
      shouldComponentUpdate (newProps, newState) {
        return (newState.intersectionNode === nodeName);
      }

      render() {
        const { props } = this;
        const component = createElement(
          RouteSegment,
          { ...props }
        );

        return component;
      }
    }

    return RouteNode;
  };
}

export default routeNode;
```

And used like this:

```
import Section from './section'
export default routeNode('section')(Sections);
```

So when the observable `route` from the `routerStore` will be equal to `section` the  HOC  will re-render its wrapped component (and so all its children).

Although this works it feels much like a hack.
Is this the correct way to accomplish a custom re-render behaviour?

Does exist (or should be created) an API to do this? I mean instead of `@observer` create a similar pattern that will re-render if a custom condition (according to us) is met, of course depending on a change of an observable.

**Comment by mweststrate** [comment]
> What if we want to fine-tune when the render should be triggered? 

There is no way to customize the tracking behavior. The whole idea is that you should not implement `shouldComponentDidUpdate` on your own. `observer` will only react to observables that are actually used in the last run of the render (so not to observables that are behind some if statement that returns falsy). Being able to customize this would allow you to break the basic guarantee of MobX: views are always in sync with state, and is not a desirable situation. 

I'm not entirely grabbing the problem from your comment, but the solution looks quite complicated :). I am not sure why the tree shouldn't rerender if some data which is apparently relevant for that subtree changes.

**Comment by LeonardoGentile** [comment]
Hello, thanks for the feedback.

> I am not sure why the tree shouldn't rerender if some data which is apparently relevant for that subtree changes.

Each RouteNode is a wrapper around the nodes responsible for rendering part of the tree. 

When transitioning from a path to another only few of all the RouteNodes needs to re-render so that's why of `return (newState.intersectionNode === nodeName);`

I'm using this to implement a plugin for router5 I'm working on, you can read more about these ideas here: http://router5.github.io/docs/understanding-router5.html

> On a route change, you only need to re-render a portion of your app. Depending on where you come from, for the same given route, a smaller or larger part of your application view will need to be re-rendered. This is why route handlers are not helpful: routing is not about mapping a route to a component, it is about going from A to B.

Now, removing the shouldComponentUpdate will work as well, but if I already know that the sub-components of a given RouteNode will never need re-rendering for for a given `intersectionNode`  why not avoid some extra computation if possible?

Maybe this computations are really cheap and I'm making the problem bigger than it is 😄

**Comment by mweststrate** [comment]
Trust me, if you need introduce a lot of complexity and it is unclear what performance you gain, you are probably just shooting yourself in the foot :). In practice it turns out to be pretty hard to outsmart / outperform mobx with these kind of things. Some interesting blog posts have been written about that.

**Comment by LeonardoGentile** [comment]
Yes, you are probably right, but I was talking more about avoiding unnecessary calls of React render methods of all the subcomponents of RouteNode when `newState.intersectionNode != nodeName` because I already know they won't produce any visual or side effect.. 🤔

**Comment by andykog** [comment]
@LeonardoGentile, doesn't using
```js
const intersectionNode = expr(() => some.newState.intersectionNode);
```
helps avoiding unnecessary re-renders? ([docs/expr](https://mobx.js.org/refguide/expr.html))

**Comment by LeonardoGentile** [comment]
@andykog I didn't know about expr, Nice! I played around with it but I think it doesn't fit for my use case. 

Maybe I was not clear, the first rendering needs to happen, what I want is to prevent the following re-rendering (what I obtained with shouldComponentUpdate) when a condition is not met.   
I've accomplished this as in my first post with a mix of `setState`, `autorun` and `shouldComponentUpdate`.   
This works because the `autorun` updates the state and AFAIK the `shouldComponentUpdate` can only control things when only state is involved, that means that it won't work if the component is an observer and there is at least one observable used inside the render.

What I really needs is to prevent a **re**-rendering. The first one should happen. Also changing any of the 2 the observables in the render will trigger another rendering anyway. So I wouldn't know how to use expr for this case... 🤔 

To recap:
1) on mount and first rendering: observable props should be forwarded
2) on observable changes: trigger a re-rendering only if the condition this.routerStore.intersectionNode === nodeName is true. If positive then forward the observable props down.

So probably I should give up or use the code in my first comment, as long as shouldComponentUpdate will be supported (I guess not for long tough..)

**Comment by andykog** [comment]
@LeonardoGentile, I still believe `expr` suites perfectly for this use case (if not, please provide some jsfiddle showing the unnecessary re-rendering).
Lets say, you are using 2 observable values `state.a` and `state.b` to determine current route. Instead of using them directly in render(), you can use `const currentRoute = expr(() => getCurrentRoute(state.a, state.b))`. Then re-rendering will only happen if the resulted `currentRoute` changed, regardless of what happened to `a` and `b`.

**Comment by robstown** [comment]
I have  a need to orchestrate renders as well.   I have an app with tens of thousands of components.  I have to be very careful removing, adding, and updating components because the react comparison takes too long for even the simplest changes.  Maybe when  @mweststrate introduces his own vdom with derived arrays this will be solved :

> from Roadmap:  introduce derived arrays, #166 (and for fun built a VDom with it 🎱)

My main focus is to maintain a snappy UI. For that reason, almost every component is an observer and I try to minimize changes to top level arrays.  The problems is that changing some global states can cause all or many components to re-render.  This can take 0.5s-2s, which makes for a sluggish experience.  To avoid this, I track what content is visible on the screen.  I update only those until the user is done changing values, and then update everything else.  

I am doing this now using complex @computed logic and additional layers in the data store.  This seems overly complicated and fragile, and each time I have a new observable I want to add, I need to implement additional logic for it.  

What would be ideal is to add a method like 'shouldObserverReact' to @observer(s) or an optional holdUpdate input variable to the observer instead.  Observables passed into this method are tracked but the render is not triggered when falsey values are returned.

As contrived example, lets say I have an array of 10,000 boxes.  I may use the following code:

```
class Box {
	ID;
	states;
	label;
	prevBox;
	constructor(ID, label, states, prevBox) {
		this.ID = ID;
		this.label = label;
		this.states = states;
		this.prevBox = prevBox;
	}
	@observable width = 10;
	@observable height = 5;
	@computed get color() {return this.states.colorScheme.boxes};
	@computed get displayWidth() {return this.width * this.states.horizontalZoom;}
	@computed get displayHeight() {return this.height * this.states.verticalZoom;}
	@computed get visible() {return this.states.filter ? this.label.indexOf(this.states.filter) >= 0 : true;}
	@computed get xStart() {return !this.prevBox ? 0 : this.prevBox.xEnd;}
	@computed get xEnd() {return this.visible ? this.xStart : this.xStart + this.displayHeight;}
	@computed get inView() {
		return (this.xStart >= this.states.scrollPos && this.xStart <= this.states.screenEndX) || (this.xEnd >= this.states.scrollPos && this.xEnd <= this.states.screenEndX);
	}
	@computed.struct get update() {this.inView || (!this.firstPass && !this.screenUpdate);}

class States {
	@observable screenUpdate = false;
	@observable firstPass = false;
	@observable verticalScale = 1;
	@observable horizontalScale = 1;
	@observable filter;
	@observable scrollPosX = 0;
	@observable screenHeight = 800;
	@observable colorScheme = {boxes:'blue', circles:'red'};
	@computed get screenEndX() {return this.scrollPosX + this.screenHeight;}
}

@observer
class DisplayBox extends React.Component {
	shouldObserverReact(nextProps){
		return nextProps.store.update && nextProps.store.visible && this.props.store.visible;
	}
	render() {
		let {store, style, ...props} = this.props;
		style = {...{
			width:store.displayWidth, 
			height:store.displayHeight, 
			backgroundColor:store.color, 
			diplay: store.visible ? 'block' : 'none';
		}, ...style};
		return (
			<div style={style} className='displayBox'>
				{store.label}
			</div>
		)
	}
}
```
In this example, let's say the user has a slider that changes the value of 'horizontalScale', which affects of all the box components.  On the mousedown of the slider, I can set 'screenUpdate' to true so while he is changing the box lengths, only the boxes on the screen update (and thus it appears to work real-time and not janky).  On mousedown I can set 'screenUpdate' back to false so the rest of the boxes update.  Likewise, if I have a button that changes the box colors I can set 'firstPass' to true and set a timer or requestAnimationFrame to change it back to false after the initial render.  Finally, any box that is not visible due to a global filter will not be updated assuming it was invisible before and after an update. 

I think the desired behavior would be to update the observer component anytime:

 1) an observable in the render function changes and shouldObserverReact returns true 
    -OR-
 2) when shouldObserverReact changes from false to true.  

Ideally, the observer-component would keep track of whether any observables in render changed while shouldObserverReact was false and only update those components when shouldObserverReact changes to true.

Do you know a pattern that exists with current Mobx implementation that would support this? 

Thanks!

**Comment by robstown** [comment]
I just started catching up on React Fiber from the React conference last month...
[https://www.youtube.com/watch?v=ZCuYPiUIONs](url)

It seems like most of my headache of reducing jank might be addressed through assigning higher priority for onscreen components in the next React Fiber version, and thus a moot point for Mobx?  This seems like it will also address @mweststrate concern for orchestrating renders using Mobx:

> Being able to customize this would allow you to break the basic guarantee of MobX: views are always in sync with state, and is not a desirable situation.

From a data standpoint the view will be in sync, and from a render standpoint, you have a guarantee it will be in sync once the processor has had time to catch up (unless you starve it).  This will be great for getting rid of transition time for hidden tabbed windows too, but may change the way you tell Mobx to react to observer components that are not visible on the screen (will have to use display:none or stacking of content instead of excluding the content from render logic).  I am looking forward to see how this pans out!

**Comment by urugator** [comment]
> Do you know a pattern that exists with current Mobx implementation that would support this?

@robstown What about preventing a subscription for other observables, when it's not visible:

```javascript
@observer
class DisplayBox extends React.Component {
  // original render method
  renderHelper() {
    let {store, style, ...props} = this.props;
    style = {...{
      width:store.displayWidth, 
      height:store.displayHeight, 
      backgroundColor:store.color, 
      diplay: store.visible ? 'block' : 'none';
    }, ...style};
   return (
     <div style={style} className='displayBox'>{store.label}</div>
    )
  }
  render() {
    // when not visible, subscribe only for store.visible
    if (this.props.store.visible === false) {  
      return untracked(() => this.renderHelper());
    }
    // otherwise, subscribe for everything
    return this.renderHelper(); 
  }
}
``` 
So when component "dissapears" it also unsubcribes from irrelevant observables.

**Comment by robstown** [comment]
@urugator , thanks!  I'll try that approach.  I guess I assumed that any observable reference behind an if statement was treated the same as one that was not.   Your saying is that I can make my own conditionally tracked function?  Technically, I guess it is opening and closing subscriptions, but that is just as good.  I didn't think passing in an untracked version would unsubscribe the original.  If this works, it will open a lot of doors for me!

**Comment by urugator** [comment]
> I guess I assumed that any observable reference behind an if statement was treated the same as one that was not.

What really matters is whether the observable has been actually accessed the last time the reaction (render) run. Accessing the observable is causing the subscription, but it's more like calling `eventListener.once` than `eventListener.on`, therefore the "unsubscription" is not very precise term - there is no need to "unsubscribe", there is only need to "subscribe again".      
If an observable is not accessed, because it's behind some if statement, which returned false, because of another observable used inside the condition, then there is no need to notify the observer, because the result of reaction (render) doesn't depend on that unused observable, unless the accessed observable inside the condition changes, causing the reaction to re-run and possibly subscribing for the osbervable behind the if statement.     
So the `untracked` is not used to "unsubscribe", but to access observables, without "subscribing" for them.

**Comment by urugator** [comment]
@robstown There is also another approach you can think about ... the idea would be that observer reacts only to changes ... so you don't want some properties to actually change when the box is not visible...that can be reflected in the model itself:
```javascript
@computed get displayWidth() {
  if (this.visible === false) {
    return 0;
  }
  return this.width * this.states.horizontalZoom;
}
```
Now, unless the box is visible, the width won't change (it's always 0), therefore no observer is going to be notified.
The problem might be that the width doesn't always reflect the actual width, so there might be some other reactions interested about height even if the box is not visible.
There is nothing stopping you to provide different API for different consumers, so you can introduce multiple `displayWidth` methods or an adapter, which does the visibility check before delegating the call to actual box model...

**Comment by urugator** [comment]
@robstown In fact you can keep the visibility check inside the reaction:
```javascript
// inside render
width: expr(() => store.visible ? store.displayWidth : 0),
```  
The effect is the same.

**Comment by robstown** [comment]
Thanks @urugator , I am doing something similar now.  I store the last used value in a non-observable property and use that value if I don't want to render instead of 0 or null.  I used a _.struct_ so there is no reaction if the same value is returned.  If I return a 0 instead, it will still cause a reaction the first time it switches from the calculated display width to the 0 value.  If a use a previous value it doesn't.  Here is a quick example:

```
class Box {
	prevUsedDisplayWidth;
	@observable width = 10;
	@computed.struct get displayWidth() {
		if(this.shouldRender){
			displayWidth = this.width * this.states.horizontalZoom;
			this.prevUsedWidth = displayWidth;
			return displayWidth;
		}
		else return this.prevUsedWidth;
	}
}
```
Using this method I have the added advantage of caching the last render value in memory so I can use it for further computed values, for example, to derive the starting scroll position of that component by adding all the previous displayWidths.  This method adds layers and complexity in the store, but I've learned to mitigate that somewhat using generic class structures.

**Comment by urugator** [comment]
>  If I return a 0 instead, it will still cause a reaction the first time it switches from the calculated display width to the 0 value.

But it will switch to this value only if the `store.visible` switches as well, so the reaction runs either way (it reacts to `store.visible` change), you won't save any renderings by doing so...

**Comment by robstown** [comment]
Oh yeah, thanks!

**Comment by robstown** [comment]
That works for switching from visible to non-visible, but I wanted a case that would also work for rendering what is in view now vs what is scrolled out of view.  For those cases, I have to use the previously used value to avoid a re-render.

**Comment by urugator** [comment]
> rendering what is in view now vs what is scrolled out of view

For some calculations or do you have two kind of components - one needs to prevent rendering only when not visible and another needs to prevent rendering only when outside of view?

**Comment by urugator** [comment]
Oh, I haven't notice .... you can't have side effects inside computed:
`this.prevUsedWidth = displayWidth;`
If `this.prevUsedWidth` is observable it won't work in strict mode. Strict mode or not you shouldn't do that... If you want to cache the value do it inside setter (action).

**Comment by robstown** [comment]
Yeah, that's why I don't make this.prevUsedWidth an observable.  I really don't need it to be reactive anyways, I just used it to store older values.  As far as calculations, I have three types of renders I like to orchestrate.  If a component is in the DOM, but not visible (display:none), I don't want to update the component at all until it becomes visible.  For components mounted in the DOM and visible, I want to modify items that are in the viewport on the first pass, then all the others afterwards to make updates look faster.  On some updates, like when the user uses a slider or knob to update scaling, I only want to update components in the viewport until the user stops using the slider/knob (mouseup), then modify the rest of the components that are out of view so they are already rendered correctly when they are scrolled into view.

**Comment by urugator** [comment]
Instead of preventing reactions, woudn't be easier to prevent (or delay) the state changes? Something like:

```javascript
// Inside the slider/knob

@action onMove() {
  // update only visible
  this.boxes.filter(box => box.isInViewport()).forEach(box => {
    box.height = someValue;
  });  
}

@action onMoveEnd() {
  // update all
  this.boxes.forEach(box => {
    box.height = someValue;
  })
}
```

**Comment by robstown** [comment]
I'd be more comfortable not having so many observable states floating around (one for each box).  I'd rather keep state in one store and let each instance derive it's own value to use for DOM renders based on that central state. 

For now, I'll probably just stick to using my method of storing previous values and only updating the computed value based on the most up-to date state if my computed render function returns true.

I'm trying to understand creating decorator functions for getters better.  Ideally, I'd like to create a generic _conditional_ decorator function that would do this in a cleaner and more composable manner.  It would return a computed variable that only updates if the supplied argument returns true.  A conditionally computed method for displayHeight might look something like this:

```
// Box should render on first pass or while operating slider only if the box is within the screen viewport
@computed get shouldRender() {
	return (!this.states.firstPass && !this.states.zoomSliderActive) || this.inView;   
}

// only update displayHeight if render is true, otherwise use the last cached value
@conditional(this.shouldRender) get displayHeight(){
	return this.box.height * this.states.VertZoom
}
```

It would be nice if this conditional computed option existed within Mobx out of the box :)  If I was better at programming and used typescript I'd issue a PR!  Drinking from the firehose now though...

**Comment by urugator** [comment]
Not sure about that `@conditional` idea, however notice that you can use the same `untracked` "trick" inside the `displayHeight`
```javascript
@computed get displayHeight() {
  return this.box.height * this.states.VertZoom;
}

@computed get styleHeight() {
  // once shouldRender turns to false, computed won't be notified about displayHeight changes so the output remains the same
  if (!this.shouldRender) { 
     return untracked(this.displayHeight);
  }
  return this.displayHeight;  
}
```  
I think you also don't need the `this.states.firstPass`, because the computed result is cached only if there is some subscriber (that is when component have been rendered <=> first pass)

**Comment by LeonardoGentile** [comment]
@andykog going back to my original question you can take a look how I use `shouldComponentUpdate` from [this code](https://github.com/LeonardoGentile/react-mobx-router5/blob/master/src/modules/routeNode.js#L49). I still couldn't figure out how to use expr to solve my problem and I believe it doesn't fit for this specific use case (instead of the `shouldComponentUpdate` that btw is working but feels like a hack)

**Comment by andykog** [comment]
@LeonardoGentile, Initially I thought that the problem is that your component should be re-rendered only when `instersectionRoute` changes, but you have to lookup other props in order to obtain this `intersactionRoute`, so you can compute intersactionRoute inside untracked or even the value of `matched`: https://jsfiddle.net/Sl1v3r/v98nz8pd/

But now I digged deeper and I see that your goal was a bit different: you have 3 observable props: route, previousRoute and intersectionNode. If intersectionNode matches nodeName, you want mobx to behave normally, but otherwise it should only observe intersectionNode, still passing the last `route` and `previousRoute` to child component. Is it right? 
Then, we have to add `untracked` into the equation: https://jsfiddle.net/Sl1v3r/xoa5zxmx/

**Comment by LeonardoGentile** [comment]
@andykog thanks for taking the time to help me 👍 

I think my problem is similar to the second solution you proposed: if intersectionNode matches nodeName then re-render the child passing also the observables `route` and `previousRoute`, otherwise just don't render and don't pass anything. 

I've tried your second solution: I apply the modified routeNode HOC to several component representing "nodes" and I want them to re-render when the transition route is the correct intersection, logically your solution seems correct! But when my transition involves going from an intersection to another it triggers an infinite loop somewhere I still couldn't identify why and where.

Probably if you you could better explain this line I could be able to find the source of the trouble

```
const isIntersection = expr(() => this.routerStore.intersectionNode === nodeName);
return isIntersection ? this.renderChild() : untracked(() => this.renderChild());
```

Here what's the difference between  `this.renderChild()` and `untracked(() => this.renderChild())`?

**Comment by andykog** [comment]
@LeonardoGentile, I can't tell why you've got infinite loop without seing code. My fiddle doesn't seem to have such issue. Maybe `whyRun()` could help determine what causes infinite reactions?

> Probably if you you could better explain this line I could be able to find the source of the trouble

Everything is pretty simple: I used `expr(() => this.routerStore.intersectionNode === nodeName)` to only track swithes of this expression's result between true/false (not to track replacing some  irrelevant intersectionNode with another irrelevant intersectionNode)
If the result becomes `false`, all the rendering happens inside `untracked()`, so anything used inside won't cause re-render.
Note that this adds one superfluous re-render after expr() result swithes from true to false. If I'd want to get rid of it, I'd stick to your current approach, only without shouldComponentUpdate, something like:
```js
componentWillMount() {
  autorun(() => {
    if (expr(() =>routerStore.intersectionNode === nodeName)) {
      this.setState({ route: routerStore.route, previousRoute:  routerStore:previousRoute })
    }
  });
}
```
or (basically, the same thing):
```js
@computed get mathes() {
  return routerStore.intersectionNode === nodeName;
}

componentWillMount() {
  autorun(() => {
    if (this.matches) {
      this.setState({ route: routerStore.route, previousRoute:  routerStore:previousRoute });
    }
  });
}
```
That should result in exactly the same behaviour you have now.

**Comment by LeonardoGentile** [comment]
@andykog thanks 💯  👍 

The infinite loop was a misconfiguration on my side. 

Both of your solution work but probably the one that solve **exactly** my problem is the last one you proposed, to note that for this to work as intended we should probably remove the `@observer` from the component and just rely on the component `state`.

By the way I was not thinking about using `@computed` outside of a "store" and in conjunction with other non-observable variables, setting my mind to mobx way isn't so immediate but I'm getting using to it 😉 

Very valuable information came out of this discussion, I think adding "how to conditional rendering" to the docs would help many other folks and avoid you guys from the team to always answer the same questions here 👍

**Comment by LeonardoGentile** [comment]
@andykog implementing the last solution (the one using setState without observer) I'm facing another issue. 

I've forked your fiddle and made an example: https://jsfiddle.net/v5s9fy42/

Basically autorun observes the `@computed get isIntersection` and updates the component state accordingly. When this happen the render method creates another component with `createElement`: 

```
render() {
  const {route} = this.state;
  return createElement(MyComponent, {...this.props,  route });
}
```
The `route` prop passed to the new `MyComponent` is still an observable. 
Inside MyComponent I could generate another component forwarding down the observable `route`. Now if this last `AnotherComponent` is an `@observer` it might produce an unwanted result: the `AnotherComponent` render method would fire before the `Router` render method that is something I have to avoid! 

I guess this is due to the way setState work, quoting from react docs
> React may delay it, and then update several components in a single pass.

So I guess this is the limitation of mixing up `setState` and `@observer` together, right?

I would probably switch back to your previous solution (the one using observer and untracked).

Is my interpretation here correct or am I doing another mistake I can't see?

**Comment by andykog** [comment]
@LeonardoGentile, I don't see how https://jsfiddle.net/v5s9fy42/ is different from my fiddle.
My last proposition was to track dependencies in autorun and call setState / forceUpdate manually. 
This way you don't need to use `@observer` as render never uses observable data, so it should have no effect. The `route` prop that you get from state can't be “observable” as it is primitive value. Even if it was an ObservableObject, `createElement(MyComponent, { route })` won't track this object's properties lookups inside parent component. If It's not clear, try reading [What does MobX react to?](https://mobx.js.org/best/react.html)

**Comment by andykog** [comment]
> React may delay it, and then update several components in a single pass.

In my latest example it shouldn't matter, as all the dependencies tracked here:
```if (expr(() =>routerStore.intersectionNode === nodeName)) {```
and here:
``` this.setState({ route: routerStore.route, previousRoute:  routerStore:previousRoute })```


this.setState({ route: routerStore.route, previousRoute:  routerStore:previousRoute })

**Comment by LeonardoGentile** [comment]
@andykog sorry I updated my comment, the link correct is this one:  https://jsfiddle.net/v5s9fy42/1/

**Comment by andykog** [comment]
@LeonardoGentile, you are using `this.setState({ route: this.state.route, });`
this doesn't make sense, it could be replaced with `this.setState({});`

To make it work, you shoud use `this.setState({ route: state.route, });` https://jsfiddle.net/Sl1v3r/v5s9fy42/2/

**Comment by LeonardoGentile** [comment]
yes, I made it very quick, now that I see I spotted a couple of errors, sorry 😆  but did you get my point? When passing the observable route down to the children, and if the child is and observer, then it could render before the Router component (that uses setState instead of observe). That is a simplified ugly example, I've found this issue on a more complete code.

My problem is not that the example didn't work (I was aware) but it's that the log in the console could come (sometimes) after the result print on screen

**Comment by andykog** [comment]
Oh, simply use this.forceUpdate() then, observer uses it under the hood

**Comment by LeonardoGentile** [comment]
To me that sounds the same as going back to `shouldComponentUpdate` (things that I will do),  it will probably avoid all the issues we have been discussing so far. I guess this use case is just too complex to be handled in the "mobx way"

**Comment by mweststrate** [comment]
Closing this issue as it is becoming quite inactive (and hard to follow). But as a general direction; what usually helps in creating more fine grained control is introducing view modules, where you can use e.g. `@computed` properties to not expose value to the UI yet. This captures the intention clearly in the state and will make code more followable then custom lifecycle hooks.
